// From `github.com/microsoft/TypeScript/blob/v5.8.2/tests/cases/compiler/binaryArithmatic4.ts`, Apache-2.0 License

var v = null | null;
//~^ ERROR: The value 'null' cannot be used here.
//~| ERROR: The value 'null' cannot be used here.