  × Unexpected token. Did you mean `{'>'}` or `&gt;`?
   ╭─[../../tests/cases/compiler/tsxButStartWithGeneric/index.tsx:1:18]
 1 │ >
   · ─
   ╰────

  × Unexpected token. Did you mean `{'>'}` or `&gt;`?
   ╭─[../../tests/cases/compiler/tsxButStartWithGeneric/index.tsx:1:18]
 1 │ >
   · ─
   ╰────

  × Identifier expected.
   ╭─[../../tests/cases/compiler/tsxButStartWithGeneric/index.tsx:1:26]
 1 │ ╭─▶ ";
 2 │ │   //~^ ERROR: Expected
 3 │ │   //~| ERROR: Expected
 4 │ │   //~| ERROR: Identifier expected.
 5 │ │   //~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
 6 │ │   //~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
 7 │ ╰─▶ 
   ╰────

  × Expected '>'.
   ╭─[../../tests/cases/compiler/tsxButStartWithGeneric/index.tsx:1:26]
 1 │ ╭─▶ ";
 2 │ │   //~^ ERROR: Expected
 3 │ │   //~| ERROR: Expected
 4 │ │   //~| ERROR: Identifier expected.
 5 │ │   //~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
 6 │ │   //~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
 7 │ ╰─▶ 
   ╰────

  × Expected ','.
   ╭─[../../tests/cases/compiler/tsxButStartWithGeneric/index.tsx:1:26]
 1 │ ╭─▶ ";
 2 │ │   //~^ ERROR: Expected
 3 │ │   //~| ERROR: Expected
 4 │ │   //~| ERROR: Identifier expected.
 5 │ │   //~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
 6 │ │   //~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
 7 │ ╰─▶ 
   ╰────
