const t = <T>() => {};"</";
//~^ ERROR: Expected '
//~| ERROR: Expected '
//~| ERROR: Unterminated string literal.
//~| ERROR: Expected corresponding JSX closing tag for 'T'.
//~| ERROR: Identifier expected.
//~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
