// From `github.com/microsoft/TypeScript/blob/v5.8.3/tests/cases/compiler/badArrayIndex.ts`, Apache-2.0 License

var results = number[];
//~^ ERROR: An element access expression should take an argument.
//~| ERROR: Cannot find name 'number'.