// From `github.com/microsoft/TypeScript/blob/v5.9.2/tests/cases/compiler/downlevelLetConst11.ts`, Apache-2.0 License

"use strict";
let
//~^ ERROR: Identifier expected. 'let' is a reserved word in strict mode.
//~| ERROR: Cannot find name 'let'.