// From `github.com/microsoft/TypeScript/blob/v5.7.2/tests/cases/compiler/incompleteDottedExpressionAtEOF.ts`, Apache-2.0 License

var p2 = window. 
//~^ ERROR: Identifier expected.
//~| ERROR: Cannot find name 'window'.