const t = <T>() => {};"</";
//~^ ERROR: Expected
//~| ERROR: Expected
//~| ERROR: Identifier expected.
//~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?
//~| ERROR: Unexpected token. Did you mean `{'>'}` or `&gt;`?

